clients/src/main/java/org/apache/kafka/clients/admin/DeleteShareGroupOffsetsResult.java

```
@@ -0,0 +1,88 @@
+ * limitations under the License.
+ */
+package org.apache.kafka.clients.admin;
+
+import org.apache.kafka.common.KafkaFuture;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.annotation.InterfaceStability;
+import org.apache.kafka.common.internals.KafkaFutureImpl;
+import org.apache.kafka.common.protocol.Errors;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * The result of the {@link Admin#deleteShareGroupOffsets(String, Set, DeleteShareGroupOffsetsOptions)} call.
+ * <p>
+ * The API of this class is evolving, see {@link Admin} for details.
+ */
+@InterfaceStability.Evolving
+public class DeleteShareGroupOffsetsResult {
+
+    private final KafkaFuture<Map<TopicPartition, Errors>> future;
+
+    DeleteShareGroupOffsetsResult(KafkaFuture<Map<TopicPartition, Errors>> future) {
+        this.future = future;
+    }
+
+    /**
+     * Return a future which succeeds only if all the deletions succeed.
+     */
+    public KafkaFuture<Void> all() {
+        return this.future.thenApply(topicPartitionErrorsMap ->  {
+            List<TopicPartition> partitionsFailed = topicPartitionErrorsMap.entrySet()
+                .stream()
+                .filter(e -> e.getValue() != Errors.NONE)
+                .map(Map.Entry::getKey)
+                .collect(Collectors.toList());
+            for (Errors error : topicPartitionErrorsMap.values()) {
+                if (error != Errors.NONE) {
+                    throw error.exception(
+                        "Failed deleting share group offsets for the following partitions: " + partitionsFailed);
+                }
+            }
+            return null;
+        });
+    }
+
+    /**
+     * Return a future which can be used to check the result for a given partition.
+     */
+    public KafkaFuture<Void> partitionResult(final TopicPartition partition) {
+        final KafkaFutureImpl<Void> result = new KafkaFutureImpl<>();
```

I think it would be preferable to throw an exception directly for the situation where the supplied partition was not in the original request, rather than returning a failed future. I expect this would only occur if the application code had a bug and a simple exception is simpler to deal with because there's no need to unwrap the exception.

---

Thanks for the review. I believe that is done in `DeleteConsumerGroupOffsetsResult`, but that was because we are passing set of topicPartitions to the `DeleteConsumerGroupOffsetsResult` constructor itself. There it was possible to check whether the given partition in `partitionResult` method was a part of the set of partitions and throw the error without even waiting for `future` to complete. But since there's no such need yet to pass the set of topic partitions, we need to wait for the `future` to complete before we can make sure that the given partition is contained in the result or not. Once inside whenComplete, I believe throwing an error will only make a difference in the completion state of `future`, not `result`. Since `result` is what is being returned from the method, directly throwing an exception would be missed. I believe the correct way would be to complete the future exceptionally, or maybe have the set of topic partitions in the class itself, so that we don't need to wait for the completion of `future`. Let me know what sounds good to you, will make the changes appropriately.